source/lac/sparse_direct.cc

```diff
@@ -928,66 +928,175 @@ SparseDirectMUMPS::initialize_matrix(const Matrix &matrix)
+            {
+              for (size_type row = 0; row < matrix.m(); ++row)
+                {
+                  for (typename Matrix::const_iterator ptr = matrix.begin(row);
+                       ptr != matrix.end(row);
+                       ++ptr)
+                    if (std::abs(ptr->value()) > 0.0 && ptr->column() >= row)
+                      {
+                        a[n_non_zero_elements]   = ptr->value();
+                        irn[n_non_zero_elements] = row + 1;
+                        jcn[n_non_zero_elements] = ptr->column() + 1;
+
+                        ++n_non_zero_elements;
+                      }
+                }
+            }
+          else
+            {
+              for (size_type row = 0; row < matrix.m(); ++row)
+                {
+                  for (typename Matrix::const_iterator ptr = matrix.begin(row);
+                       ptr != matrix.end(row);
+                       ++ptr)
+                    if (std::abs(ptr->value()) > 0.0)
+                      {
+                        a[n_non_zero_elements]   = ptr->value();
+                        irn[n_non_zero_elements] = row + 1;
+                        jcn[n_non_zero_elements] = ptr->column() + 1;
+                        ++n_non_zero_elements;
+                      }
+                }
+            }
+          id.n   = n;
+          id.nnz = n_non_zero_elements;
+          id.irn = irn.get();
+          id.jcn = jcn.get();
+          id.a   = a.get();
+        }
+    }
+  else if constexpr (std::is_same_v<TrilinosWrappers::SparseMatrix, Matrix> ||
+                     std::is_same_v<PETScWrappers::MPI::SparseMatrix, Matrix>)
+    {
+      // Distributed matrix case
+      id.icntl[17]               = 3; // distributed matrix assembly
+      id.nnz                     = matrix.n_nonzero_elements();
+      nnz                        = id.nnz;
+      size_type n_non_zero_local = 0;
 
-      size_type n_non_zero_elements = 0;
+      // Get the range of rows owned by this process
+      row_range                 = matrix.locally_owned_range_indices();
+      size_type local_non_zeros = 0;
+
+      if constexpr (std::is_same_v<TrilinosWrappers::SparseMatrix, Matrix>)
+        {
+          const auto &trilinos_matrix = matrix.trilinos_matrix();
+          local_non_zeros             = trilinos_matrix.NumMyNonzeros();
+        }
+      else if constexpr (std::is_same_v<PETScWrappers::MPI::SparseMatrix,
+                                        Matrix>)
+        {
+          Mat &petsc_matrix =
+            const_cast<PETScWrappers::MPI::SparseMatrix &>(matrix)
+              .petsc_matrix();
+          MatInfo info;
+          MatGetInfo(petsc_matrix, MAT_LOCAL, &info);
+          local_non_zeros = (PetscInt)info.nz_used;
+        }
+
+
+      irn = std::make_unique<MUMPS_INT[]>(local_non_zeros);
+      jcn = std::make_unique<MUMPS_INT[]>(local_non_zeros);
+      a   = std::make_unique<double[]>(local_non_zeros);
```

Comment:
If I see this right, you allocate enough entries for the nonsymmetric case, but you may end up using fewer if the users specifies that the matrix is symmetric. Is this correct? If so, perhaps add a comment to that effect.

(One could *shrink* the array but in practice the memory consumption of the decomposition is so vastly larger than that of the original matrix that it perhaps does not matter very much.)